[Inductor][Triton][FP8] Support deepseek-style scaling in Inductor

jananisriram opened the pull request:
Summary:
Support deepseek-style scaling in Inductor Triton for FP8 GEMMs. DeepSeek-style scaling is a colloquial term for a fine-grained mixed precision framework using FP8 to train [Deepseek-V3](https://arxiv.org/pdf/2412.19437), DeepSeek AI's recent MoE (Mixture of Experts) model. DeepSeek-style scaling effectively extends the dynamic range of FP8 by mitigating dequantization overhead under increased-precision accumulation, which is key to achieving more accurate FP8 GEMM results.

DeepSeek-style scaling on matmul `A @ B` leverages two different types of scaling strategies to preserve a balance between numerical stability and training efficiency:
- Activations (input tensor `A`): tile-wise (1x128 across shape `(M, K)`)
- Weights (input tensor `B`): block-wise (128x128 across shape `(N, K)`)

This diff enables Inductor users to replicate past successes with deepseek-style scaling and achieve higher numerical stability while increasing training efficiency.

NOTE: Block-wise 128x128 scaling is only supported in CUDA 12.9+; therefore, deepseek-style scaling is currently unsupported in `fbcode` (CUDA 12.4). Use OSS PyTorch to run deepseek-style scaling.

NOTE: Accuracy for FP8 is unstable, even with high tolerances, which is why TritonBench benchmarks are unlikely to be accurate against a `torch` implementation.

Test Plan:
In OSS PyTorch, run
```
TORCHINDUCTOR_CACHE_DIR=~/personal/cache_dir_inductor CUDA_LAUNCH_BLOCKING=1 TORCH_USE_CUDA_DSA=1 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 ENABLE_PERSISTENT_TMA_MATMUL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 python run.py --op fp8_gemm --only torch_fp8_gemm,pt2_fp8_gemm --metrics tflops,accuracy --m 4096 --n 768 --k 512 --output="{output_dir}/deepseek_bench.csv" --scaling_deepseek --atol=1e-2 --rtol=0.5 2>&1 | tee ~/personal/deepseek_style/deepseek_bench.log
```

Differential Revision: D83609850


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -558,10 +559,172 @@ def apply_scaling(
+    a_desc_ptr = workspace_base
+    b_desc_ptr = workspace_base + TMA_SIZE
+
+    triton.language.extra.cuda.experimental_device_tensormap_create2d(
```
do we need experimental tma and the tensor descriptor? isn't the later the way to do this?

jananisriram replied:
i left this in because the other scaled_mm device_tma template also supports it, but if we don't want to support the experimental tma api, we can definitely remove this! @drisspg are we interested in supporting it here?

drisspg replied:
I think that is planning to be removed AFAIK, and that the tensor descriptor will be the main path forward so I think we should remove

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -407,7 +408,7 @@ def apply_scaling(
 
 
-device_tma = r"""
+scaled_mm_device_tma_tensor_row = r"""
```
maybe `row` and `tensor` right?

drisspg replied:
actually lets call this epilogue scaling

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -558,10 +559,136 @@ def apply_scaling(
 
-scaled_mm_device_tma_template = TritonTemplate(
-    name="scaled_mm_device_tma",
+scaled_mm_device_tma_tensor_row_template = TritonTemplate(
```
can we call this main_loop scaling

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1338,8 +1471,60 @@ def tuned_scaled_mm(
+                    V.graph.sizevars.statically_known_equals(d, 1) for d in sz
+                )
+
+            def _ceil_div(v: int, d: int) -> int:
```
we gotta have one of these lying around already lol

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1338,8 +1471,60 @@ def tuned_scaled_mm(
+            def _is_blockwise_1x128(
+                t: torch.Tensor, scale_t_size: torch.Tensor
+            ) -> torch.Tensor:
+                return scale_t_size[0] == t.shape[0] and scale_t_size[1] == _ceil_div(
```
did we fuzz w/ dynamic shapes to make sure we dont get any symint errors

drisspg commented on torch/_inductor/kernel/mm.py:
```diff
@@ -1253,6 +1380,12 @@ def tuned_sparse_semi_structured_mm(
+class ScalingMode(IntEnum):
+    TENSOR = 0
+    ROW = 1
+    DEEPSEEK = 2
```
name change, should align w/ aten

drisspg reviewed (commented):
Overall looks good a couple comments,

cc @slayton58

drisspg reviewed (commented):
Also, tests?

jananisriram commented on torch/_inductor/kernel/mm.py:
```diff
@@ -558,10 +559,136 @@ def apply_scaling(
+    source=scaled_mm_device_tma_tensor_row + load_scales + apply_scaling,
+)
+
+scaled_mm_device_tma_deepseek_block_group = r"""
```
also aten right now doesn't support group-wise (it's more so tile-wise), so block-wise seems to be more of an encompassing name

drisspg replied:
Yeah honestly naming is hard, I have used
PerTensor -> 1 scale
PerRow -> [M,K] Tensor  = M scales
PerGroup -> [M, K] Tensor = [M, K//group] scales with a Block_size [ 1, group]
PerBlock -> [M, K] Tensor = [M//group_a, K//group_b] scales with a Block_size [ group_a, group_b]

So group is 1d spanning block is 2d spanning

drisspg replied:
ahh okay yeah lets stick with that sorry

slayton58 reviewed (commented):
Looks good, couple of minor comments

jananisriram committed:
[Inductor][Triton][FP8] Support deepseek-style scaling in Inductor (#164404)

Summary:

Support deepseek-style scaling in Inductor Triton for FP8 GEMMs. DeepSeek-style scaling is a colloquial term for a fine-grained mixed precision framework using FP8 to train [Deepseek-V3](https://arxiv.org/pdf/2412.19437), DeepSeek AI's recent MoE (Mixture of Experts) model. DeepSeek-style scaling effectively extends the dynamic range of FP8 by mitigating dequantization overhead under increased-precision accumulation, which is key to achieving more accurate FP8 GEMM results.

DeepSeek-style scaling on matmul `A @ B` leverages two different types of scaling strategies to preserve a balance between numerical stability and training efficiency:
- Activations (input tensor `A`): tile-wise (1x128 across shape `(M, K)`)
- Weights (input tensor `B`): block-wise (128x128 across shape `(N, K)`)

This diff enables Inductor users to replicate past successes with deepseek-style scaling and achieve higher numerical stability while increasing training efficiency.

NOTE: Block-wise 128x128 scaling is only supported in CUDA 12.9+; therefore, deepseek-style scaling is currently unsupported in `fbcode` (CUDA 12.4). Use OSS PyTorch to run deepseek-style scaling.

NOTE: Accuracy for FP8 is unstable, even with high tolerances, which is why TritonBench benchmarks are unlikely to be accurate against a `torch` implementation.

Test Plan:
Note that this command only works with the benchmark command (follow-up PR): in OSS PyTorch, run
```
TORCHINDUCTOR_CACHE_DIR=~/personal/cache_dir_inductor CUDA_LAUNCH_BLOCKING=1 TORCH_USE_CUDA_DSA=1 TRITON_PRINT_AUTOTUNING=1 TRITON_ALWAYS_COMPILE=1 TORCH_LOGS=+inductor TORCHINDUCTOR_FORCE_DISABLE_CACHES=1 ENABLE_PERSISTENT_TMA_MATMUL=1 TORCHINDUCTOR_MAX_AUTOTUNE_GEMM=1 python run.py --op fp8_gemm --only torch_fp8_gemm,pt2_fp8_gemm --metrics tflops,accuracy --m 4096 --n 768 --k 512 --output="{output_dir}/deepseek_bench.csv" --scaling-pair=BlockWise1x128,BlockWise128x128 --atol=1e-2 --rtol=0.5 2>&1 | tee ~/personal/deepseek_style/deepseek_bench.log
```

Reviewed By: slayton58

Differential Revision: D83609850
